[Screen Time] Unexpected power usage due to ScreenTimeWebExtension performing CA Commits when invisible

jesxilin opened the pull request:
#### d10f7ce53734b9bb35b3b3783ddf0f19b5559eed
<pre>
[Screen Time] Unexpected power usage due to ScreenTimeWebExtension performing CA Commits when invisible
<a href="https://bugs.webkit.org/show_bug.cgi?id=299634">https://bugs.webkit.org/show_bug.cgi?id=299634</a>
<a href="https://rdar.apple.com/156520177">rdar://156520177</a>

Reviewed by Tim Horton and Aditya Keerthi.

ScreenTimeWebExtension is performing more CA commits when
not rendering, than expected. As a result, there is more power usage
than anticipated.

As a temporary solution, only install the STWebpageController for
managed-users. This is achieved by using STScreenTimeConfigurationObserver
to observe the enforcesChildRestrictions property.

* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView _screenTimeConfigurationObserver]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Change all API tests to set enforcesChildRestrictions to YES after the
WKWebView is created but before _installScreenTimeWebpageControllerIfNeeded
is called. In a follow-up PR, I will be adding a KVO API test for
enforcesChildRestrictions.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(swizzleEnforcesChildRestrictions):
(testSuppressUsageRecordingWithDataStore):
(testShowsSystemScreenTimeBlockingView):
(testWebContentIsNotClickableShowingSystemScreenTimeBlockingView):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
(TEST(ScreenTime, IdentifierNil)):
(TEST(ScreenTime, IdentifierString)):
(TEST(ScreenTime, IdentifierStringWithRemoveData)):
(TEST(ScreenTime, WKWebViewFillsStackView)):
(TEST(ScreenTime, URLIsPlayingVideo)):
(TEST(ScreenTime, URLIsPictureInPicture)):
(TEST(ScreenTime, FetchData)):
(TEST(ScreenTime, RemoveDataWithTimeInterval)):
(TEST(ScreenTime, RemoveData)):
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOccludedWebView)):
(TEST(ScreenTime, CreateControllerAfterOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):
(TEST(ScreenTime, ScreenTimeControllerViewOnlyInstalledForHTTPFamily)):
(TEST(ScreenTime, ScreenTimeControllerNotInstalledForNoChildRestrictions)):

Canonical link: <a href="https://commits.webkit.org/300778@main">https://commits.webkit.org/300778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e899fb6146a2f60e21b70022eabb0d70f5610bed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130590 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75960 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d7bf1d90-8509-4c02-a0a7-ab053cebc4be) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125687 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52119 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94158 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2c9c05cb-fbf0-4b36-b362-a1de17fd2be8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110761 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74758 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/80674f4a-8fd8-4b6d-8ec9-1e697c4deefd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34212 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28922 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74072 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104983 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29145 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133285 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50762 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38655 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102629 "5 flakes 53 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51137 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102458 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47810 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26063 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47612 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19484 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50616 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56380 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50090 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53436 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51764 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5eba9d4f7f1543acf3a0ac343648e958371a1ac7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51412

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6ac930fcc6f938662ad1dbe4935bfb155d8e979c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51412

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/13bf353ff023f2ce3a45d2f4b136aecd88ca3813. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51412

aproskuryakov commented:
If I understand this correctly, nothing changes for accounts with ScreenTime enabled, is that correct? If so, is this enough for even a temporary solution?

lilyspiniolas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -451,11 +461,18 @@ - (void)_uninstallScreenTimeWebpageController
     if (!PAL::isScreenTimeFrameworkAvailable())
         return;
 
+    if (!_screenTimeConfigurationObserver)
+        return;
+
     if (!_screenTimeWebpageController)
         return;
```
nit: maybe combine these two if statements
```suggestion
    if (!_screenTimeConfigurationObserver || !_screenTimeWebpageController)
        return;
```

hortont424 commented:
> If I understand this correctly, nothing changes for accounts with ScreenTime enabled, is that correct? If so, is this enough for even a temporary solution?

It avoids the overhead for everyone without child restrictions (not ScreenTime alone).

jesxilin commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -451,11 +461,18 @@ - (void)_uninstallScreenTimeWebpageController
 
     [std::exchange(_screenTimeBlurredSnapshot, nil) removeFromSuperview];
 
+    [_screenTimeConfigurationObserver stopObserving];
```
Oh, I think it will. I can reorder the logic to only create `_screenTimeConfigurationObserver` if all other conditions are satisfied or add the code to `stopObserving` before the `_screenTimeWebpageController` check.

jesxilin replied:
Actually the first way is impossible since we need to have a `_screenTimeConfigurationObserver` to check `enforcesChildRestrictions`, then install _screenTimeWebpageController or not. So, I will go with the second way.

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -417,6 +418,12 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+    if (!_screenTimeConfigurationObserver) {
+        _screenTimeConfigurationObserver = adoptNS([PAL::allocSTScreenTimeConfigurationObserverInstance() initWithUpdateQueue:globalDispatchQueueSingleton(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)]);
+        [_screenTimeConfigurationObserver addObserver:self forKeyPath:@"configuration.enforcesChildRestrictions" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&screenTimeConfigurationObserverKVOContext];
+        [_screenTimeConfigurationObserver startObserving];
```
Another problem: we do this work even if ST framework is not available, and even if the pref is disabled. I don't think we should

jesxilin replied:
Will be moving the observer code to after `if (!_page->preferences().screenTimeEnabled() || !_page->mainFrame()->url().protocolIsInHTTPFamily())`

aproskuryakov commented:
> It avoids the overhead for everyone without child restrictions (not ScreenTime alone).

I think that those with child restrictions also tend to be those with older devices that are more in need of a fix here.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d8fd12ee3c75d9a96d1c859d150bbd5d4dfdc57d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51412

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -173,13 +187,17 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
     RetainPtr webView = webViewForScreenTimeTests(configuration.get());
 
     RetainPtr request = [NSURLRequest requestWithURL:[NSURL URLWithString:@"http://webkit.org"]];
+    auto swizzle = waitAndSetEnforcesChildRestrictions(childRestrictionsDone);
```
a bit odd that the function says "wait" but the waiting is actually outside of it.

jesxilin replied:
Will change to maybe just "setEnforcesCHildRestrictions" or 'forceEnforcedChildRestrictions"?

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -423,7 +424,16 @@ - (void)_installScreenTimeWebpageControllerIfNeeded
+
+    if (!_screenTimeConfigurationObserver) {
+        _screenTimeConfigurationObserver = adoptNS([PAL::allocSTScreenTimeConfigurationObserverInstance() initWithUpdateQueue:globalDispatchQueueSingleton(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)]);
+        [_screenTimeConfigurationObserver addObserver:self forKeyPath:@"configuration.enforcesChildRestrictions" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&screenTimeConfigurationObserverKVOContext];
```
You do not appear to use either old nor new values from the observation dict so I think you can get rid of those

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -480,10 +497,8 @@ - (void)_updateScreenTimeBasedOnWindowVisibility
         [self _installScreenTimeWebpageControllerIfNeeded];
-        if (!previouslyInstalledScreenTimeWebpageController && _screenTimeWebpageController)
-            [_screenTimeWebpageController setURL:[self _mainFrameURL]];
+        [_screenTimeWebpageController setURL:[self _mainFrameURL]];
```
I don't quite follow the reasoning for the old logic; make sure it wasn't important?

pxlcoder reviewed (commented):
Ideally we'd also have a test that exercises the configuration change via KVO.

Typo in commit message: "that anticipated" -> "than anticipated".

jesxilin commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -436,7 +488,7 @@ static void testWebContentIsNotClickableShowingSystemScreenTimeBlockingView(bool
     EXPECT_FALSE(blurredViewIsPresent(webView.get()));
 }
 
-TEST(ScreenTime, WKWebViewFillsStackView)
+TEST(ScreenTime, DISABLED_WKWebViewFillsStackView)
```
I will make the view onscreen instead

jesxilin commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -510,6 +525,13 @@ - (void)_updateScreenTimeBasedOnWindowVisibility
 {
 #if ENABLE(SCREEN_TIME)
+    if (context == &screenTimeConfigurationObserverKVOContext) {
+        ensureOnMainRunLoop([self] {
+            [self _updateScreenTimeBasedOnWindowVisibility];
```
ooh, gotcha!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -423,7 +424,16 @@ - (void)_installScreenTimeWebpageControllerIfNeeded
         return;
 
-    if (!_page->preferences().screenTimeEnabled() || !_page->mainFrame()->url().protocolIsInHTTPFamily())
+    if (!_page->preferences().screenTimeEnabled() || !_page->mainFrame() || _page->mainFrame()->url().isNull() || !_page->mainFrame()->url().protocolIsInHTTPFamily())
```
I expect that just `!_page->mainFrame() ||` is sufficient to fix what you were seeing. It's `_page->mainFrame()->url().isNull()` that is likely unnecessary.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -592,6 +587,7 @@ struct PerWebProcessState {
 - (RefPtr<WebKit::WebPageProxy>)_protectedPage;
 #if ENABLE(SCREEN_TIME)
 - (STWebpageController *)_screenTimeWebpageController;
+- (STScreenTimeConfigurationObserver *)_screenTimeConfigurationObserver;
```
Ditto.

pxlcoder reviewed (approved):
We can add the KVO test separately.

jesxilin committed:
[Screen Time] Unexpected power usage due to ScreenTimeWebExtension performing CA Commits when invisible
https://bugs.webkit.org/show_bug.cgi?id=299634
rdar://156520177

Reviewed by Tim Horton and Aditya Keerthi.

ScreenTimeWebExtension is performing more CA commits when
not rendering, than expected. As a result, there is more power usage
than anticipated.

As a temporary solution, only install the STWebpageController for
managed-users. This is achieved by using STScreenTimeConfigurationObserver
to observe the enforcesChildRestrictions property.

* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView _screenTimeConfigurationObserver]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Change all API tests to set enforcesChildRestrictions to YES after the
WKWebView is created but before _installScreenTimeWebpageControllerIfNeeded
is called. In a follow-up PR, I will be adding a KVO API test for
enforcesChildRestrictions.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(swizzleEnforcesChildRestrictions):
(testSuppressUsageRecordingWithDataStore):
(testShowsSystemScreenTimeBlockingView):
(testWebContentIsNotClickableShowingSystemScreenTimeBlockingView):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
(TEST(ScreenTime, IdentifierNil)):
(TEST(ScreenTime, IdentifierString)):
(TEST(ScreenTime, IdentifierStringWithRemoveData)):
(TEST(ScreenTime, WKWebViewFillsStackView)):
(TEST(ScreenTime, URLIsPlayingVideo)):
(TEST(ScreenTime, URLIsPictureInPicture)):
(TEST(ScreenTime, FetchData)):
(TEST(ScreenTime, RemoveDataWithTimeInterval)):
(TEST(ScreenTime, RemoveData)):
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOccludedWebView)):
(TEST(ScreenTime, CreateControllerAfterOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):
(TEST(ScreenTime, ScreenTimeControllerViewOnlyInstalledForHTTPFamily)):
(TEST(ScreenTime, ScreenTimeControllerNotInstalledForNoChildRestrictions)):

Canonical link: https://commits.webkit.org/300778@main
